File: src/lattices/hexagonal.jl
#   Copyright 2023 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
"""
Functions that support computations specific to hexagonal lattices
"""
# --- Exports

# Types
export HexagonalLatticeConstants

# --- Types

"""
    HexagonalLatticeConstants

Lattice constants for a hexagonal unit cell

Fields
======
* `a`, `c`: lengths of the edges of the unit cell

Supertype: [`LatticeConstants`](@ref)
"""
struct HexagonalLatticeConstants <: LatticeConstants
    # Fields
    a::Float64
    c::Float64

    """
    Construct a set of hexagonal lattice constants.
    """
    function HexagonalLatticeConstants(a::Real, c::Real)

        # --- Enforce constraints

        if a <= 0
            throw(ArgumentError("`a` must be positive"))
        end

        if c <= 0
            throw(ArgumentError("`c` must be positive"))
        end

        # --- Construct and return new HexagonalLatticeConstants

        return new(a, c)
    end
end

# --- Functions/Methods

# ------ LatticeConstants functions

function isapprox(
    x::HexagonalLatticeConstants,
    y::HexagonalLatticeConstants;
    atol::Real=0,
    rtol::Real=atol > 0 ? 0 : √eps(),
)
    return isapprox(x.a, y.a; atol=atol, rtol=rtol) &&
           isapprox(x.c, y.c; atol=atol, rtol=rtol)
end

function lattice_system(::HexagonalLatticeConstants)
    return Hexagonal()
end

# ------ Unit cell computations

function basis(lattice_constants::HexagonalLatticeConstants)
    # Get lattice constants
    a = lattice_constants.a
    c = lattice_constants.c

    # Construct basis
    basis_a = Vector{Float64}([a, 0, 0])
    basis_b = Vector{Float64}([-0.5 * a, SIN_PI_OVER_THREE * a, 0])
    basis_c = Vector{Float64}([0, 0, c])

    return basis_a, basis_b, basis_c
end

function volume(lattice_constants::HexagonalLatticeConstants)
    # Get lattice constants
    a = lattice_constants.a
    c = lattice_constants.c

    # Compute volume
    return a^2 * SIN_PI_OVER_THREE * c
end

function surface_area(lattice_constants::HexagonalLatticeConstants)
    # Get lattice constants
    a = lattice_constants.a
    c = lattice_constants.c

    # Compute surface area
    return 4 * a * c + 2 * a^2 * SIN_PI_OVER_THREE
end

File: pysrc/xtallography/unit_cell/hexagonal.py
#   Copyright 2025 Velexi Corporation
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
"""
The hexagonal module defines classes and methods specific to hexagonal lattices.
"""

# --- Imports

# Standard library
import copy
from typing import Optional

# Local packages/modules
from xtallography.symmetry import Centering, LatticeSystem

from .. import _JL
from .lattice_constants import HexagonalLatticeConstants
from .unit_cell import UnitCell
from .unit_cell_symmetry import UnitCellSymmetry


# --- Classes


class HexagonalUnitCell(UnitCell):
    """
    Class representing a hexagonal unit cell
    """

    # --- Initializer

    def __init__(
        self,
        a: float,
        c: float,
        centering: Optional[Centering] = Centering.PRIMITIVE,
        symmetry_elements: Optional[set] = copy.deepcopy(set()),
    ):
        """
        Initialize HexagonalUnitCell object.

        Parameters
        ----------
        `a`, `c`: lattice constants

        `centering`: centering of unit cell

        `symmetry_elements`: symmetry elements that unit cell is invariant under
        """
        # --- Check arguments

        if a <= 0:
            raise ValueError(f"`a` must be positive. (a={a})")

        if c <= 0:
            raise ValueError(f"`c` must be positive. (c={c})")

        # --- Initialize parent class

        lattice_constants = HexagonalLatticeConstants(a, c)
        symmetry = UnitCellSymmetry(centering, symmetry_elements)
        super().__init__(LatticeSystem.HEXAGONAL, lattice_constants, symmetry)

    # --- Properties

    @property
    def a(self):
        """
        Return `a`.
        """
        return self.lattice_constants.a

    @property
    def c(self):
        """
        Return `c`.
        """
        return self.lattice_constants.c

    # --- Methods

    def to_julia(self):
        """
        Convert HexagonalUnitCell object to a Julia UnitCell object.
        """
        return _JL.HexagonalUnitCell(
            self.a,
            self.c,
            centering=self.centering.to_julia(),
            symmetry_elements=_JL.Vector(
                [element.to_julia() for element in self.symmetry_elements]
            ),
        )

    @classmethod
    def from_julia(cls, unit_cell_jl: _JL.UnitCell):
        """
        Convert a Julia UnitCell object to a HexagonalUnitCell object.
        """
        # --- Check arguments

        if not _JL.isa(unit_cell_jl, _JL.HexagonalUnitCell):
            raise ValueError(
                "`unit_cell_jl` must be a Julia `HexagonalUnitCell` object. "
                f"(unit_cell_jl={unit_cell_jl})."
            )

        # --- Convert unit_cell_jl to a HexagonalUnitCell object

        unit_cell_symmetry = UnitCellSymmetry.from_julia(unit_cell_jl.symmetry)
        unit_cell = HexagonalUnitCell(
            unit_cell_jl.lattice_constants.a,
            unit_cell_jl.lattice_constants.c,
            centering=unit_cell_symmetry.centering,
            symmetry_elements=unit_cell_symmetry.symmetry_elements,
        )

        return unit_cell

    def __repr__(self):
        """
        Return string representation of HexagonalUnitCell.
        """
        return (
            f"HexagonalUnitCell(a={self.a},c={self.c},"
            f"centering={self.centering},"
            "symmetry_elements=["
            f"{','.join(sorted([str(item) for item in self.symmetry_elements]))}"
            "])"
        )
